fdbserver/worker.actor.cpp

```diff
@@ -748,17 +749,61 @@ ACTOR Future<Void> workerSnapCreate(WorkerSnapRequest snapReq, StringRef snapFol
 	return Void();
 }
 
-ACTOR Future<Void> monitorServerDBInfo( Reference<AsyncVar<Optional<ClusterControllerFullInterface>>> ccInterface, Reference<ClusterConnectionFile> connFile, LocalityData locality, Reference<AsyncVar<ServerDBInfo>> dbInfo ) {
+ACTOR Future<Void> monitorTraceLogIssues(Optional<Reference<AsyncVar<std::set<std::string>>>> issues) {
+	state bool pingTimeout = false;
+	loop {
+		wait(delay(SERVER_KNOBS->TRACE_LOG_FLUSH_FAILURE_CHECK_INTERVAL_SECONDS));
+		ThreadFuture<Void> f(new ThreadSingleAssignmentVar<Void>);
+		Reference<CompletionCallback<Void>> callback =
+		    Reference<CompletionCallback<Void>>(new CompletionCallback<Void>(f));
+		callback->self = callback;
+		f.callOrSetAsCallback(callback.getPtr(), callback->userParam, 0);
+		pingTraceLogWriterThread(f);
+		try {
+			wait(timeoutError(callback->promise.getFuture(), SERVER_KNOBS->TRACE_LOG_PING_TIMEOUT_SECONDS));
+		} catch (Error& e) {
+			if (e.code() == error_code_timed_out) {
+				pingTimeout = true;
+			} else {
+				throw;
+			}
+		}
+		if (issues.present()) {
+			std::set<std::string> _issues = getTraceLogIssues();
+			if (pingTimeout) {
+				// Ping trace log writer thread timeout.
+				_issues.insert("trace_log_writer_thread_unresponsive");
+				pingTimeout = false;
+			}
+			if (_issues.size() > 0) {
+				issues.get()->set(_issues);
```

Comment:
I'm not sure if we should think about this now or not, but if we ever intend for this issues variable to be modifiable in multiple places, we would need for this to use insertion rather than assignment. It would need to be done carefully such that we remove items that are no longer relevant.

I bring this up because it's used in a somewhat generic place (`monitorServerDBInfo`) and someone could conceivably see it and try to use it.

Reply:
That's a good point. I added a TODO in the code as both a note for the caller of the function and also an action item for later.